server/src/main/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticator.java

```diff
@@ -235,10 +248,21 @@ private boolean isSubjectAuthorized(final ClientDetails expectedClient, final St
     }
 
     private void assertAudience(final Map<String, Object> claims, final String issuerURL) {
-        String audience = (String) claims.get(ClaimConstants.AUD);
-
-        if (StringUtils.isEmpty(audience) || !audience.equals(issuerURL)) {
-            throw new InvalidTokenException("Audience does not match.");
+        Object audienceObject = claims.get(ClaimConstants.AUD);
+        if(audienceObject instanceof ArrayList) {
+            ArrayList<?> audienceList = (ArrayList<?>) audienceObject;
+            if(!audienceList.contains(issuerURL)) {
+                throw new InvalidTokenException("Audience does not match.");
+            }
+        }
+        else if(audienceObject instanceof String) {
+            String audience = (String) audienceObject;
+            if (StringUtils.isEmpty(audience) || !audience.equals(issuerURL)) {
+                throw new InvalidTokenException("Audience does not match.");
+            }
+        }
+        else {
+            throw new InvalidTokenException("aud claim is in the wrong format.");
```

Comment:
This is unreachable code in case of **invalid** audience claim.

Reply:
Yes this is unreachable as I felt weird having an else if block without an else. I can remove it.

Reply:
Although, this safeguards us if we (for whatever reason) remove validateClaims call before the assertAudience call. As a standalone function it would be successful if the wrong input type was passed.

Reply:
Pardon me. My point was not about else block, which makes sense to be there for wrong input type. But to the entire rechecking of the audience claim in assertAudienc. As in the case of invalid Audience the code doing the same thing never gets called.

Reply:
This is checking something different though... it is checking whether the audience matches the issuer url which is different than checking that the value exists. Both are needed and a value for audience must exist in order to check whether it matches the issuer url. If the comment is not about the else block then I am missing your point.